// cords-handler/src/classes/EventHandler/index.js
const fs = require("fs");
class CordsHandler {
  constructor(client, chclient) {
    if (!chclient)
      throw new ReferenceError("Cords-Handler Client Must Be Supplied.");
    if (!client) throw new ReferenceError("Cannot Find Client.");
    else {
      this.client = client;
      this.chclient = chclient;
    }
    return this;
  }

  setEventFolder(eventfolder) {
    if (!eventfolder) throw new ReferenceError("Cannot Find Event Folder.");
    else {
      this.eventfolder = eventfolder;
    }
    return this;
  }
  loader() {
    let events = fs.readdirSync(`${process.cwd()}/${this.eventfolder}`);
    for (let event of events) {
      const evnt = require(`${process.cwd()}/${this.eventfolder}/${event}`);
      var eventName = event.split(".")[0];
      if (eventName === "ready") {
        this.client.on("ready", async () => evnt(this.client));
      } else {
        this.client.on(eventName, (...args) => evnt(...args));
      }
    }
    this.chclient.emit("eventsLoaded");
    return this;
  }
}
module.exports = CordsHandler;
