// cords-handler/src/structures/ReactionRole/index.js
const discord = require("discord.js");
class ReactionRole {
  constructor(client) {
    if (!client) throw new ReferenceError("Cannot Find Client.");
    else {
      this.client = client;
    }
  }
  setMessage(msg) {
    if (!msg) throw new ReferenceError("Cannot Find Message.");
    else {
      this.msg = msg;
    }
    return this;
  }
  setEmoji(emoji) {
    if (!emoji) throw new ReferenceError("Cannot Find Emoji/Emojis.");
    else {
      this.emoji = emoji;
    }
    return this;
  }
  setRole(role) {
    if (!role) throw new ReferenceError("Cannot Find Role.");
    else {
      this.role = role;
    }
    return this;
  }
  start() {
    let msg = this.msg;
    let emoji = this.emoji;
    let role = this.role;
    let client = this.client;
    msg.react(emoji);
    client.on("messageReactionAdd", async (reaction, user) => {
      if (user.bot) return;
      if (!reaction.message.guild) return;
      if (reaction.message.partial) await reaction.message.fetch();
      if (reaction.partial) await reaction.fetch();
      if (!reaction.message.guild.id === msg.guild.id) return;
      if (!reaction.message.id === msg.id) return;
      if (reaction.emoji.name === emoji) {
        await reaction.message.guild.members.cache.get(user.id).roles.add(role);
      }
    });
    client.on("messageReactionRemove", async (reaction, user) => {
      if (user.bot) return;
      if (!reaction.message.guild) return;
      if (reaction.message.partial) await reaction.message.fetch();
      if (reaction.partial) await reaction.fetch();
      if (!reaction.message.guild.id === msg.guild.id) return;
      if (!reaction.message.id === msg.id) return;
      if (reaction.emoji.name === emoji) {
        await reaction.message.guild.members.cache
          .get(user.id)
          .roles.remove(role);
      }
    });
  }
}
module.exports = ReactionRole;
