// cords-handler/src/structures/AutoReply/index.js
class AutoReply {
  constructor(client) {
    if (!client) throw new ReferenceError("Cannot Find Client.");
    else {
      this.client = client;
    }
  }
  setReplyMessage(msg) {
    if (!msg) throw new ReferenceError("Cannot Find Reply Message.");
    else {
      this.replymsg = msg;
    }
    return this;
  }
  setTargetMessage(msg) {
    if (!msg) throw new ReferenceError("Cannot Find Target Message.");
    else {
      this.targetmsg = msg;
    }
    return this;
  }

  setOptions(options) {
    if (!options) throw new ReferenceError("Please Use Options.");
    if (!options === "mention" && options === "lowerupper")
      throw new ReferenceError("This Is Not An Valid Option.");
    else {
      this.options = options;
    }
    return this;
  }
  start() {
    this.client.on("message", async (msg) => {
      if (this.options.mention === true) {
        if (!this.options.lowerupper === true) {
          if (msg.content === this.targetmsg) return msg.reply(this.replymsg);
        } else {
          if (msg.content.toLowerCase() === this.targetmsg)
            return msg.reply(this.replymsg);
        }
      }
      if (!this.options.mention === true) {
        if (!this.options.lowerupper === true) {
          if (msg.content === this.targetmsg)
            return msg.channel.send(this.replymsg);
        } else {
          if (msg.content.toLowerCase() === this.targetmsg)
            return msg.channel.send(this.replymsg);
        }
      }
    });
    return this;
  }
}
module.exports = AutoReply;
