// cords-handler/src/structures/WelcomeLeave/index.js
const Discord = require("discord.js");
class WelcomeLeave {
  constructor(client, chclient) {
    if (!client) throw new ReferenceError("Cannot Find Client.");
    if (!chclient)
      throw new ReferenceError("Cords-Handler Client Must Be Supplied.");
    else {
      this.chclient = chclient;
      this.client = client;
    }
  }
  setChannel(id) {
    if (!id) throw new ReferenceError("Cannot Find Channel ID.");
    else {
      this.chid = id;
    }
    return this;
  }
  setRole(id) {
    if (!id) throw new ReferenceError("Cannot Find Role ID.");
    else {
      this.rlid = id;
    }
    return this;
  }
  setMessage(msg) {
    if (!msg) throw new ReferenceError("Cannot Find Message.");
    else {
      this.msg = msg;
    }
    return this;
  }
  start() {
    this.client.on("guildMemberAdd", async (member) => {
      if (this.rlid) {
        member.roles.add(this.rlid);
      }
      let msg1 = this.msg;
      let message = msg1
        .replace("{mention}", `<@${member.id}>`)
        .replace("{guildname}", `${member.guild.name}`)
        .replace("{membername}", `${member.user.name}`)
        .replace("{membertag}", `${member.user.tag}`);
      this.client.channels.cache.get(this.chid).send(message);
      this.chclient.emit("welcomeMessageSent");
    });

    return this;
  }
}
module.exports = WelcomeLeave;
